### ISSUE TITLE
Multiple output regression

### ISSUE THREAD

**Issue by miguelmartin75** [dmlc/xgboost#2087]
How do I perform multiple output regression? Or is it simply not possible?

My current assumption is that I would have to modify the code-base such that XGMatrix supports a matrix as labels and that I would have to create a custom objective function.

My end goal would be to perform regression to output two variables (a point) and to optimise euclidean loss. Would I be better off to make two seperate models (one for x coordinates and one for y coordinates).

Or... would I be better off using a random forest regressor within sklearn or some other alternative algorithm?

**Comment by khotilov** [comment]
Multivariate/multilabel regression is not currently implemented #574 #680
Tianqi had added some relevant placeholder data structures to gbtree learner, but no one had time, I guess, to work the machinery out.

**Comment by jindongwang** [comment]
Pity, since many competitions are with multi-outputs

**Comment by MarkusBonsch** [comment]
This would be a really nice feature to have.

**Comment by joel-thomas-wilson** [comment]
Do we have any updates on this?

**Comment by JacobKempster** [comment]
I agree - this feature would be extremely valuable (exactly what I need right now...)

**Comment by cp9612** [comment]
Would like to see this feature coming

**Comment by veonua** [comment]
any reason why it is closed?

**Comment by hcho3** [comment]
@veonua See #3439.

**Comment by loretoparisi** [comment]
In the meanwhile there is any alternative, like any ensemble of single output models?

**Comment by jimmywan** [comment]
> In the meanwhile there is any alternative, like any ensemble of single output models like:

https://scikit-learn.org/stable/modules/generated/sklearn.multioutput.MultiOutputRegressor.html

**Comment by cmottet** [comment]
I am going to also weigh and say that having such feature would be extremely handy. The MultiOutputRegressor mentioned above is a nice wrapper to build multiple models at once and it does work well for predicting target variables that are independent from one another. However, if the target variables are highly correlated, then you really want to build one model that predicts a vector.

**Comment by MxNl** [comment]
A year has passed soon since the last comment :-). This is why I want to repeat the wish to have such an interesting feature. I would be happy to see this. Thanks anyway for all your work.

**Comment by hcho3** [comment]
Reopening for visibility.

**Comment by kk26269** [comment]
> Multivariate/multilabel regression is not currently implemented #574 #680
> Tianqi had added some relevant placeholder data structures to gbtree learner, but no one had time, I guess, to work the machinery out.

Hello ,I have used the SckitLearn estimator and passed my script (.py)written for multioutput regression to the same and I could create endpoints.
I have reffered following repo.
https://github.com/qlanners/ml_deploy/tree/master/aws/scikit-learn/sklearn_estimators_locally.
Changes done are :
 
Y= dataset.iloc[:,-3:]
X = dataset.iloc[:,:-3]
X_train, X_test , Y_train , Y_test =train_test_split(X, Y , test_size = 0.20,random_state =100)
#     
gbr = GradientBoostingRegressor()
modelMOR = MultiOutputRegressor(estimator=gbr)
modelMOR.fit(X_train, Y_train)

**Comment by mirik123** [comment]
The `MultiOutputRegressor` is bad alternative because it doesn't update eval_set dataset together with the main train (X, y) dataset.

**Comment by trivialfis** [comment]
I would love to spend some time on this ...

**Comment by loretoparisi** [comment]
> I would love to spend some time on this ...

I have used this approach and it seems to work fine

https://github.com/dmlc/xgboost/issues/2087#issuecomment-534640535

**Comment by StatMixedML** [comment]
Is there any update on this? Can we make it a joint effort to have the multioutput regression available. Irrespective of the independence modelling of several responses/y-variables, it would be great to have the `xgb.DMatrix` accept a list or a np.array with shape >1.

**Comment by trivialfis** [comment]
Note to myself:  We should consider including the possibility of having independent early stopping for each target.

**Comment by trivialfis** [comment]
The initial support is merged in https://github.com/dmlc/xgboost/pull/7514 .  The feature is still quite primitive at the moment and is considered to be experimental.  Thank you to everyone who participated in the thread.

**Comment by giorizzi** [comment]
> The initial support is merged in #7514 . The feature is still quite primitive at the moment and is considered to be experimental. Thank you to everyone who participated in the thread.

I am trying to use a custom objective with the multiple output regressor.
Could you comment about the input and output shapes the custom objective?
The following seems to work in `demo/guide-python/multioutput_regression.py`:
```
def pseudo_huber_error(y_true, y_pred):
    y_true = y_true.reshape(y_pred.shape)
    z = y_pred - y_true
    scale = 1 + z**2
    scale_sqrt = np.sqrt(scale)
    grad = z / scale_sqrt
    hess = 1 / (scale * scale_sqrt)
    return grad.flatten(), hess.flatten()
```
therefore, `y_true.shape=(200,)` , `y_pred.shape=(100, 2)` and `grad.flatten().shape=(200,)`
Is that correct?

**Comment by trivialfis** [comment]
Let me add the parameter in pseudo huber instead.

**Comment by trivialfis** [comment]
Related https://github.com/dmlc/xgboost/issues/4840

**Comment by StatMixedML** [comment]
Based on the above discussion, I have extended the univariate [XGBoostLSS](https://github.com/StatMixedML/XGBoostLSS) to a multivariate one called [Multi-Target XGBoostLSS Regression](https://arxiv.org/abs/2210.06831) that models multiple targets and their dependencies in a probabilistic regression setting. Code follows soon. Let me know your comments.

**Comment by trivialfis** [comment]
Thank you for sharing! Would love to read the paper this weekend.

**Comment by StatMixedML** [comment]
@trivialfis Thanks for making the multi-output available in the fist place!

Would be interested in your feedback, especially on how to improve the runtime for high-dimensional responses. Problem is the know scaling issue of XGBoost for multi-class and multi-output responses, since for each target, a separate tree is grown.

**Comment by trivialfis** [comment]
> Can we change the way xgboost is trained?

Yes. I made it work with the exact and approx tree method (hist is very similar) in my prototype branches.  I will focus on approx and hist in the future. One problem with approx (and hist) is that the histogram we build needs to account for all targets. Consider a histogram with 256 bins and a 64x64 image as both input and output (encoder-decoder alike), it will have 64^4 * 256 bins. For a small number of targets, this is perfectly fine. However, once the number of targets goes up we will have challenges in training the model efficiently.

I think there's a paper for addressing the issue by defining a different type of gradient to approximate the gain function further. The new type of gradient is a form of a weighted sum of gradients from all targets. I can't recall the name of that paper off top of my head, @jameslamb might have better insight since there was a wip PR for lightgbm for implementation by the authors. It's not a perfect solution (correct me if I'm wrong, haven't really dived into it yet, apologies), since in the end we still need to calculate the full gradient for the leaf value and the approximation doesn't seem to have any error bound, but it's a very good start for investigating the issue.

**Comment by StatMixedML** [comment]
> For a small number of targets, this is perfectly fine. However, once the number of targets goes up we will have challenges in training the model efficiently.

It would be enormously beneficial for xgboost training if we can decrease the computational time it takes for high-dimensional classification / regression tasks, both univariate and multivariate. There were already some issues on this repo (https://github.com/dmlc/xgboost/issues/2926 or https://github.com/dmlc/xgboost/issues/4079), so there is a need for this. As an example for the probabilistic multivariate case, say we have `D=30` responses and we assume the data can be described by a Normal distribution. If you want to model all means, variances and dependencies between all targets, you'd need to estimate `D(D+3)/2=495` parameters. Given the size of the dataset, this might be above and beyond any reasonable runtime.

> I started reading your paper today, the review of others' work is very extensive and I'm sure there is a lot of catching up I need to do to follow up on the latest development. Will come back to this once I can get the 1.7 out. Multi-target and probabilistic forecasting are exciting topics I would love to learn more about.

Sure take your time. Let me know in case you need some more explanations when it comes to probabilistic modelling.

**Comment by jameslamb** [comment]
👋 @trivialfis thanks for the `@`

Are you maybe thinking of `MT-GBM` (where `MT` is "multi-target")?

* issue: https://github.com/microsoft/LightGBM/issues/5047
* draft PR that its authors abandoned: https://github.com/microsoft/LightGBM/pull/5111
* paper: "MT-GBM: A Multi-Task Gradient Boosting Machine with Shared Decision Trees" (https://arxiv.org/abs/2201.06239)

According to the authors:

> *"The MT-GBM can find the shared tree structures and split branches according to multi-task losses. First, it assigns multiple outputs to each leaf node. Next, it computes the gradient corresponding to each output (task)."*

**Comment by StatMixedML** [comment]
@jameslamb 

Thanks for sharing the paper. I have read the paper some time ago, so I am not fully in the details. The authors propose a multi-**task** version of LightGBM, so I am not sure this is the same as multi-**target**. I would need to go through the paper again.

In case this would be an interesting extension, do you have some resources to work on this?

**Comment by jameslamb** [comment]
To be honest, I also am not sure whether it's exactly the same. But it should be similar, right? Whether you are working on multiple **tasks** like "regression and classification" or multiple **targets** like "regression predicting y_1 and y_2", you still are in a situation like "find splits that balance gain across multiple loss functions".

To be honest, I haven't read the paper and am not planning to actively work on this (we have many other higher priorities in LightGBM right now).

**Comment by StatMixedML** [comment]
@jameslamb 

> To be honest, I haven't read the paper and am not planning to actively work on this (we have many other higher priorities in LightGBM right now).

Sure, I understand that. I am not sure I find the time either. So maybe let's pause this and see if the community picks it up.

**Comment by StatMixedML** [comment]
There is a new Boosting Approach that is designed for efficient training in multi-output classification and regression tasks

[SketchBoost: Fast Gradient Boosted Decision Tree for Multioutput Problems](https://openreview.net/forum?id=WSxarC8t-T)](https://github.com/sb-ai-lab/Py-Boost). 

It is accepted as a Paper at Neurips-2022. The following shows a runtime-comparison.

![image](https://user-images.githubusercontent.com/41187941/203947900-d00b0e36-f31c-48e3-aeaf-e9a5143824d6.png)

**Comment by trivialfis** [comment]
Did a quick scan over a couple of papers. I don't have a good understanding of various algorithms yet, but vector leaf seems to be the essential component of all proposed methods. I will try to prioritize it and share a roadmap for a path forward.

**Comment by StatMixedML** [comment]
@trivialfis Ok nice! Looking forward

**Comment by StatMixedML** [comment]
@trivialfis This might be an interesting approach to incorporate into XGBoost

[SketchBoost: Fast Gradient Boosted Decision Tree for Multioutput Problems ](https://openreview.net/forum?id=WSxarC8t-T)

The paper says

> Moreover, the proposed methods are easy to implement upon modern boosting frameworks such as XGBoost

You find the code here https://github.com/sb-ai-lab/Py-Boost

**Comment by trivialfis** [comment]
@StatMixedML Thank you for the references. Here's an early version of vector-leaf: https://github.com/dmlc/xgboost/pull/8616 No specific optimization yet.

**Comment by StatMixedML** [comment]
@trivialfis Very nice, I`ll have a look into it

**Comment by trivialfis** [comment]
I will cleanup the code in the coming days. There are some known issues that break the existing code, the only thing that's working is the demo at the moment. It's for discussion and far from ready.

**Comment by StatMixedML** [comment]
@trivialfis Sure, take your time. Let me know once I can use it.

Looking forward to it!

**Comment by StatMixedML** [comment]
@trivialfis I have seen that you created a PR for a first version of the multi-target tree. This is awesome!!

Let me know once I can test it. Would be great to run some examples and compare accuracy and runtime. Willing to volunteer on this!

**Comment by trivialfis** [comment]
Hi @StatMixedML @lcrmorin, thank you for volunteering! The PR is not ready yet, I still need to figure out some parts of the parameter interface and do more tests. If you really want to try to code, the `demo/guide-python/multioutput_regression.py` would be a good starting place, See the `rmse_model` function and parameters used in there.

**Comment by StatMixedML** [comment]
@lcrmorin So the advantage of using multi-output models is that you don't have to train a separate model for each response variable. Also, as outlined in  [Multi-Target XGBoostLSS Regression](https://arxiv.org/abs/2210.06831), you can model dependencies between the different responses. What @trivialfis is currently working on is to speed-up the estimation using multi-target trees, since currently, XGBoost trains a separate model for each response.

**Comment by loretoparisi** [comment]
Wow! Shared
https://twitter.com/loretoparisi/status/1623958275279519745?s=46&t=bN2GaGENwqbkg0nK-MY7tg

**Comment by lcrmorin** [comment]
So that would also help inference time, right ?

**Comment by StatMixedML** [comment]
@lcrmorin We would expect to see the highest efficiency gains during training time, especially for HPO / cross-validation.

**Comment by trivialfis** [comment]
The bare bone implementation is merged, please help test out. :-)
You can find reference for nightly build in xgboost's Python installation document. The computation performance is not yet optimized, please expect some quirks.

**Comment by trivialfis** [comment]
The link will be available once the CI passes

**Comment by trivialfis** [comment]
A bug fix PR for prediction along with a small optimization: https://github.com/dmlc/xgboost/pull/8968 .

**Comment by lcrmorin** [comment]
Just being curious; would this allows some missing / masked targets ? (I have multi-targets time-series applications in mind, where longer horizons targets are not available immediatly).

**Comment by trivialfis** [comment]
Not planned at the moment, the label is required to be dense. But I will mark that as a feature request and see if we can find a way to train boosting tree models with missing labels.

**Comment by trivialfis** [comment]
Hi all, thank you for joining the discussion and the helpful feedbacks! Let's continue the discussion in https://github.com/dmlc/xgboost/issues/9043 .
